File: app/src/main/java/com/locatible/qrcodereader/MainActivity.java
package com.locatible.qrcodereader;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import com.scandit.barcodepicker.ScanSettings;
import com.scandit.recognition.Barcode;
import com.scandit.recognition.BarcodeScanner;
import com.scandit.recognition.BarcodeScannerSession;
import com.scandit.recognition.ImageDescription;
import com.scandit.recognition.RecognitionContext;

import org.bytedeco.javacv.FFmpegFrameGrabber;
import org.bytedeco.javacv.Frame;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.ByteBuffer;

import static org.bytedeco.javacpp.avutil.AV_PIX_FMT_ARGB;
import static org.bytedeco.javacpp.avutil.AV_PIX_FMT_BGR24;
import static org.bytedeco.javacpp.avutil.AV_PIX_FMT_GRAY8;
import static org.bytedeco.javacpp.avutil.AV_PIX_FMT_RGB24;
import static org.bytedeco.javacpp.avutil.AV_PIX_FMT_RGBA;


public class MainActivity extends AppCompatActivity {

    private static final String VideoUrl = "rtsp://192.168.0.80";
    private static final String ScanditSdkAppKey = "AVIMQwRFGcoxJgGUhzv6j6Yo1jX3NNeL5kbO6AFUAIxDfr/GjndhnkgfnNLxeemxOWtKwgZ3hTFTAy2TP2yBH/0JHZ8FaMKLnwW/4DheP4riS5XsQxBHAJkwqfBgCzB2GUJiJCulop8Av1dV4wef9ZG+dHiBUUA9jyVlnw2aBYD/wXK2PkOsu76jkzHeiNODQgsvUvAoXKRz9PpspZ6jzbgzGn6azOSToDTT4gD9sqleZGIO69/m4vls8i4j66svDM7m3gZjG2x03SwPBUICQw7q38wu5ZSgHi0XbPQfUdpadH+AsAY/EZI9F1DcUvm4Y+dV6PyiZoVwI93VtOJtBAe56xFhoBFWleeo0nPJB38NqNNUnX7BgI/Zc5p2od/lzJKp0OltFgiHH+5FzhDsMF5Tx6TjCqrGzxOETGumCHwPGLXautnXTWEtXY5k0zds6ZH4I/sJ/WOuX1SSdvj7tOQKDpVAYNm3AB0rPo0XGfvfxbpmJoQCO+gpzfulwvyaTNBiCQNHwhkJbEQS2SXhmsGlDkq4CArg4/VvoUyxt4jzw/ewwuxNudNhQ9veRDp37mCnl35DGKcJvWB20qHO4EDTw1YLgRoH/tSgrmGin7JgS8SHTNYYigc71I2lloXCN1XZvbjU92DSfpl56fiesZ12DaMuKkrv2pxiGDE6y3yNHywrj7jpkSflUDq3Y97kr7fLpOOOe5epiqmv00w2kT1bgRVkaxySjkuKptcjwd8C9agegxT7ShKiTjxG/7YEiSCi7ecXCshzVg/B65O8KnmFkVSSAzElAtAf//9UXyHtHA==";

    private Thread videoProcessorThread;
    private Runnable videoProcessor = new Runnable() {
        @Override
        public void run() {
            try {
                processVideo();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        init();
    }

    private void init() {
        videoProcessorThread = new Thread(videoProcessor, "processVideo");
        videoProcessorThread.start();
    }

    private void processVideo() throws IOException {
        BarcodeScanner scanner = createScanner();
        RecognitionContext context = scanner.getContext();
        BarcodeScannerSession session = scanner.getSession();

        try (FFmpegFrameGrabber video = openVideo(VideoUrl)) {
            video.setOption("rtsp_transport", "tcp");
            video.start();
            System.out.println("Streaming...");

            ImageDescription imageDescription = createImageDescription(video);
            byte[] frameArray = new byte[0];

            for (; ; ) {
                Frame frame = video.grabImage();

                if (frame == null) {
                    System.out.println("Done");
                    break;
                }

                if (frame.image == null || frame.image.length == 0) {
                    continue;
                }

                ByteBuffer frameBuffer = (ByteBuffer) frame.image[0];
                frameBuffer.position(0);
                int size = frameBuffer.remaining();

                if (size == 0) {
                    continue;
                }

                if (size > frameArray.length) {
                    frameArray = new byte[size * 2];
                }

                frameBuffer.get(frameArray, 0, size);
                imageDescription.setMemorySize(size);

                context.processFrame(imageDescription, frameArray);

                for (Barcode newlyRecognizedCode : session.getNewlyRecognizedCodes()) {
                    System.out.println(newlyRecognizedCode.getData());
                }
            }
        }
    }

    private BarcodeScanner createScanner() {
        ScanSettings settings = ScanSettings.create();
        settings.setSymbologyEnabled(Barcode.SYMBOLOGY_QR, true);

        RecognitionContext context = RecognitionContext.create(this, ScanditSdkAppKey, getNoBackupFilesDir());
        context.startNewFrameSequence();

        return BarcodeScanner.create(context, settings.getBarcodeScannerSettings());
    }

    private FFmpegFrameGrabber openVideo(String url) throws IOException {
        if (url.startsWith("rtsp:")) {
            return new FFmpegFrameGrabber(url);

        } else if (url.startsWith("http")) {
            InputStream inputStream = new URL(url).openConnection().getInputStream();
            return new FFmpegFrameGrabber(inputStream);
        }

        return new FFmpegFrameGrabber(url);
    }

    private ImageDescription createImageDescription(FFmpegFrameGrabber video) {
        ImageDescription imageDescription = new ImageDescription();
        imageDescription.setWidth(video.getImageWidth());
        imageDescription.setHeight(video.getImageHeight());
        imageDescription.setLayout(getLayout(video.getPixelFormat()));
        return imageDescription;
    }

    private int getLayout(int pixelFormat) {
        switch (pixelFormat) {
            case AV_PIX_FMT_GRAY8:
                return ImageDescription.IMAGE_LAYOUT_GRAY_8U;
            case AV_PIX_FMT_RGB24:
            case AV_PIX_FMT_BGR24:
                return ImageDescription.IMAGE_LAYOUT_RGB_8U;
            case AV_PIX_FMT_RGBA:
                return ImageDescription.IMAGE_LAYOUT_RGBA_8U;
            case AV_PIX_FMT_ARGB:
                return ImageDescription.IMAGE_LAYOUT_ARGB_8U;
            default:
                return ImageDescription.IMAGE_LAYOUT_UNKNOWN;
        }
    }
}

File: app/src/main/java/com/locatible/qrcodereader/MainActivity.kt
package com.locatible.qrcodereader

import android.content.Context
import android.databinding.BaseObservable
import android.databinding.Bindable
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.locatible.qrcodereader.databinding.ActivityMainBinding
import com.scandit.barcodepicker.ScanSettings
import com.scandit.recognition.*
import org.bytedeco.javacpp.avutil.*
import org.bytedeco.javacv.FFmpegFrameGrabber
import org.bytedeco.javacv.Frame
import java.net.URL
import java.nio.ByteBuffer

class MainActivity : AppCompatActivity() {

    internal var model: MainActivityViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        init()
    }

    private fun init() {
        model = MainActivityViewModel(this)

        val binding: ActivityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        binding.model = model
    }
}

internal class MainActivityViewModel(private val activity: MainActivity) : BaseObservable() {

    private val pref = activity.getPreferences(Context.MODE_PRIVATE)
    private var videoProcessor: VideoProcessor? = null
    private var videoProcessorThread: Thread? = null

    @get:Bindable
    var videoUrl: String = pref.getString(activity.getString(R.string.pref_video_url_key), activity.getString(R.string.pref_video_url_default))
        set(value) {
            field = value
            notifyPropertyChanged(BR.videoUrl)
            with(pref.edit()) {
                putString(activity.getString(R.string.pref_video_url_key), value)
                apply()
            }
        }

    @get:Bindable
    var scanditLicense: String = pref.getString(activity.getString(R.string.pref_scandit_license_key), activity.getString(R.string.pref_scandit_license_default))
        set(value) {
            field = value
            notifyPropertyChanged(BR.scanditLicense)
            with(pref.edit()) {
                putString(activity.getString(R.string.pref_scandit_license_key), value)
                apply()
            }
        }

    @get:Bindable
    var status: String = ""
        set(value) {
            field = value
            notifyPropertyChanged(BR.status)
        }

    @get:Bindable
    var recognizedCode: String = ""
        set(value) {
            field = value
            notifyPropertyChanged(BR.recognizedCode)
        }

    @get:Bindable
    var running: Boolean = false
        set(value) {
            field = value
            notifyPropertyChanged(BR.running)
        }

    fun reset() {
        videoUrl = activity.getString(R.string.pref_video_url_default)
        scanditLicense = activity.getString(R.string.pref_scandit_license_default)
    }

    fun toggle() {
        if (running) {
            running = false
            status = activity.getString(R.string.stopping)
            if (videoProcessor != null) {
                videoProcessor!!.stop()
                videoProcessor = null
                videoProcessorThread = null
            }
        } else {
            running = true
            status = activity.getString(R.string.starting)
            videoProcessor = VideoProcessor(activity)
            videoProcessorThread = Thread(videoProcessor, "VideoProcessor")
            videoProcessorThread!!.start()
        }
    }
}

internal class VideoProcessor(private val activity: MainActivity) : Runnable {

    private val duplicateFilterDurationMs = 5000
    private val sessionClearCycle = 0xfff // Must be 2^x-1

    private val model: MainActivityViewModel = activity.model!!
    private var scanner: BarcodeScanner
    private var recognitionContext: RecognitionContext
    private var scannerSession: BarcodeScannerSession
    private var imageDescription = ImageDescription()
    private var frameData: ByteArray
    private var running = true

    init {
        val settings = ScanSettings.create()
        settings.setSymbologyEnabled(Barcode.SYMBOLOGY_QR, true)
        settings.codeDuplicateFilter = duplicateFilterDurationMs

        recognitionContext = RecognitionContext.create(activity, model.scanditLicense, activity.noBackupFilesDir)
        recognitionContext.startNewFrameSequence()

        scanner = BarcodeScanner.create(recognitionContext, settings.barcodeScannerSettings)

        scannerSession = scanner.session

        frameData = ByteArray(0)
    }

    override fun run() {
        try {
            processVideo()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun stop() {
        running = false
    }

    private fun processVideo() {
        while (running) {
            openVideo(model.videoUrl).use { video ->
                video.setOption("rtsp_transport", "tcp")
                video.start()

                setImageDescription(video)
                model.status = activity.getString(R.string.streaming)
                var cycle = 0

                while (running) {
                    // Clear session periodically to avoid memory overflow
                    if (cycle and sessionClearCycle == 0) {
                        scannerSession.clear()
                    }

                    val frame = video.grabImage() ?: break

                    if (updateFrame(frame)) {
                        recognitionContext.processFrame(imageDescription, frameData)

                        for (barcode in scannerSession.newlyRecognizedCodes) {
                            println(barcode.data)
                        }

                        if (scannerSession.newlyRecognizedCodes.size > 0) {
                            model.recognizedCode = scannerSession.newlyRecognizedCodes.last().data
                        }
                    }

                    cycle++
                }

                model.status = activity.getString(R.string.done)
            }
        }

        model.status = activity.getString(R.string.stopped)
    }

    private fun openVideo(url: String): FFmpegFrameGrabber {
        if (url.startsWith("rtsp:")) {
            return FFmpegFrameGrabber(url)

        } else if (url.startsWith("http")) {
            val inputStream = URL(url).openConnection().getInputStream()
            return FFmpegFrameGrabber(inputStream)
        }

        return FFmpegFrameGrabber(url)
    }

    private fun setImageDescription(video: FFmpegFrameGrabber) {
        imageDescription.width = video.imageWidth
        imageDescription.height = video.imageHeight
        imageDescription.layout = getLayout(video.pixelFormat)
    }

    private fun getLayout(pixelFormat: Int): Int {
        return when (pixelFormat) {
            AV_PIX_FMT_GRAY8 -> ImageDescription.IMAGE_LAYOUT_GRAY_8U
            AV_PIX_FMT_RGB24, AV_PIX_FMT_BGR24 -> ImageDescription.IMAGE_LAYOUT_RGB_8U
            AV_PIX_FMT_RGBA -> ImageDescription.IMAGE_LAYOUT_RGBA_8U
            AV_PIX_FMT_ARGB -> ImageDescription.IMAGE_LAYOUT_ARGB_8U
            else -> ImageDescription.IMAGE_LAYOUT_UNKNOWN
        }
    }

    private fun updateFrame(frame: Frame): Boolean {
        if (frame.image == null || frame.image.isEmpty()) {
            return false
        }

        val frameBuffer = frame.image[0] as ByteBuffer
        frameBuffer.position(0)
        val size = frameBuffer.remaining()

        if (size == 0) {
            return false
        }

        if (size > frameData.size) {
            frameData = ByteArray(size * 2)
        }

        frameBuffer.get(frameData, 0, size)
        imageDescription.setMemorySize(size)

        return true
    }
}
